cpp_tests/test_Test.cpp

```diff
@@ -0,0 +1,417 @@
+	remote->salEventSub((char *)"Test_logevent_arrays");
+	controller->salEventPub((char *)"Test_logevent_arrays");
+
+	getNewestAfterGetOldest<Test_logevent_arraysC>(
+		std::bind(&SAL_Test::logEvent_arrays, controller, _1, 1),
+		std::bind(&SAL_Test::getNextSample_logevent_arrays, remote, _1),
+		std::bind(&SAL_Test::getSample_logevent_arrays, remote, _1));
+	getNewestAfterGetOldest<Test_logevent_arraysC>(
+		std::bind(&SAL_Test::logEvent_arrays, controller, _1, 1),
+		std::bind(&SAL_Test::getSample_logevent_arrays, remote, _1),
+		std::bind(&SAL_Test::getSample_logevent_arrays, remote, _1), true);
+    }
+
+    /**
+     * Test that get newest after get oldest gets the newest value.
+     * <p>
+     * This tests DM-18491.
+     */
+    SECTION("TelGetNewestAfterGetOldest") {
+	remote->salTelemetrySub((char *)"Test_arrays");
+	controller->salTelemetryPub((char *)"Test_arrays");
+
+	getNewestAfterGetOldest<Test_arraysC>(std::bind(&SAL_Test::putSample_arrays, controller, _1),
+					      std::bind(&SAL_Test::getNextSample_arrays, remote, _1),
+					      std::bind(&SAL_Test::getSample_arrays, remote, _1));
+    }
+
+    SECTION("TelGetNewestAfterGetOldest") {
+	remote->salTelemetrySub((char *)"Test_arrays");
+	controller->salTelemetryPub((char *)"Test_arrays");
+
+	getNewestAfterGetOldest<Test_arraysC>(std::bind(&SAL_Test::putSample_arrays, controller, _1),
+					      std::bind(&SAL_Test::getNextSample_arrays, remote, _1),
+					      std::bind(&SAL_Test::getSample_arrays, remote, _1));
+    }
+
+    /**
+     * Test that a late joiner can can read the most recent event using getNextSample.
+     * <p>
+     * Only one value is retrievable.
+     */
+    SECTION("EvtLateJoinerOldest") {
+	controller->salEventPub((char *)"Test_logevent_arrays");
+
+	Test_logevent_arraysC dataList[maxLoops];
+	for (int i = 0; i < maxLoops; i++) {
+	    Test_logevent_arraysC data;
+	    fillArraysWithRandomValues<Test_logevent_arraysC>(&data);
+	    dataList[i] = data;
+	    REQUIRE(controller->logEvent_arrays(&data, 1) == SAL__OK);
+	}
+
+	Test_logevent_arraysC data;
+	remote->salEventSub((char *)"Test_logevent_arrays");
+
+	REQUIRE(remote->getNextSample_logevent_arrays(&data) == SAL__OK);
+	arraysEqual<Test_logevent_arraysC>(&data, &(dataList[0]));
+    }
+
+    /**
+     * Test that a late joiner cannot see historical telemetry using getNextSample.
+     * <p>
+     * Telemetry is volatile so there should be no late joiner data.
+     */
+    SECTION("TelLateJoinerOldest") {
+	controller->salTelemetryPub((char *)"Test_arrays");
+
+	for (int i = 0; i < maxLoops; i++) {
+	    Test_arraysC data;
+	    fillArraysWithRandomValues<Test_arraysC>(&data);
+	    REQUIRE(controller->putSample_arrays(&data) == SAL__OK);
+	}
+
+	Test_arraysC data;
+	remote->salTelemetrySub((char *)"Test_arrays");
+	REQUIRE(remote->getNextSample_arrays(&data) == SAL__NO_UPDATES);
+    }
+
+    /**
+     * Test that a late joiner can see an event using getEvent.
+     */
```

Comment:
getEvent will get all data. Please consider improving this to check that.

Reply:
Re-implemnted.